server/src/main/java/org/apache/cloudstack/vm/lease/VMLeaseManagerImpl.java

```diff
@@ -0,0 +1,302 @@
+import org.apache.cloudstack.framework.jobs.impl.AsyncJobVO;
+import org.apache.cloudstack.managed.context.ManagedContextTimerTask;
+import org.apache.commons.lang3.time.DateUtils;
+
+import javax.inject.Inject;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Timer;
+import java.util.TimerTask;
+
+import static com.cloud.vm.VirtualMachine.State.Destroyed;
+import static com.cloud.vm.VirtualMachine.State.Expunging;
+import static com.cloud.vm.VirtualMachine.State.Unknown;
+
+public class VMLeaseManagerImpl extends ManagerBase implements VMLeaseManager, Configurable {
+
+
+    public static ConfigKey<Boolean> InstanceLeaseEnabled = new ConfigKey<>(ConfigKey.CATEGORY_ADVANCED, Boolean.class,
+            "instance.lease.enabled", "false", "Indicates whether to enable the Instance Lease feature",
+            true, List.of(ConfigKey.Scope.Global));
+
+    private static final int ACQUIRE_GLOBAL_LOCK_TIMEOUT_FOR_COOPERATION = 5;   // 5 seconds
+
+    @Inject
+    private UserVmJoinDao userVmJoinDao;
+
+    @Inject
+    private AlertManager alertManager;
+
+    @Inject
+    private AsyncJobManager asyncJobManager;
+
+    private AsyncJobDispatcher asyncJobDispatcher;
+
+    Timer vmLeaseTimer;
+
+    Timer vmLeaseAlterTimer;
+
+    @Override
+    public String getConfigComponentName() {
+        return VMLeaseManager.class.getSimpleName();
+    }
+
+    @Override
+    public ConfigKey<?>[] getConfigKeys() {
+        return new ConfigKey[]{
+                InstanceLeaseEnabled,
+                InstanceLeaseDuration,
+                InstanceLeaseExpiryAction,
+                InstanceLeaseSchedulerInterval,
+                InstanceLeaseAlertSchedule,
+                InstanceLeaseAlertStartsAt
+        };
+    }
+
+    public void setAsyncJobDispatcher(final AsyncJobDispatcher dispatcher) {
+        asyncJobDispatcher = dispatcher;
+    }
+
+    @Override
+    public boolean start() {
+        final TimerTask schedulerPollTask = new ManagedContextTimerTask() {
+            @Override
+            protected void runInContext() {
+                try {
+                    poll(new Date());
+                } catch (final Throwable t) {
+                    logger.warn("Catch throwable in VM lease scheduler ", t);
+                }
+            }
+        };
+
+        final TimerTask leaseAlterSchedulerTask = new ManagedContextTimerTask() {
+            @Override
+            protected void runInContext() {
+                try {
+                    alert();
+                } catch (final Throwable t) {
+                    logger.warn("Catch throwable in VM lease scheduler ", t);
+                }
+            }
+        };
+
+        vmLeaseTimer = new Timer("VMLeasePollTask");
+        vmLeaseTimer.scheduleAtFixedRate(schedulerPollTask, 5_000L, InstanceLeaseSchedulerInterval.value() * 1000L);
```

Comment:
poll interval should be very frequent, shouldn't be configurable (if any limit for the value configured? if high interval value is set, lease action performed lately).